src/v/kafka/server/server.cc

```diff
@@ -288,10 +288,18 @@ ss::future<> server::apply(ss::lw_shared_ptr<net::connection> conn) {
         .kafka_throughput_controlled_api_keys.bind<std::vector<bool>>(
           &convert_api_names_to_key_bitmap));
 
+    std::exception_ptr eptr;
     try {
+        co_await ctx->start();
         co_await ctx->process();
     } catch (...) {
-        auto eptr = std::current_exception();
+        eptr = std::current_exception();
+    }
+    if (!eptr) {
+        co_return co_await ctx->stop();
```

Comment:
@rockwotj did you tell me about an upstream seastar utility for dealing with this pattern?

Reply:
ss::coroutine::as_future

```cpp
auto fut = co_await ss::coroutine::as_future(ctx->start())
if (!fut.failed()) {
  fut = co_await ss::coro::as_future(ctx->process());
}
if (fut.failed()) {
  co_return co_await ctx->stop();
}
```

Reply:
ahhh right, thanks for the reminder.

Reply:
Nice(r).